tcmur_device.h

```diff
@@ -32,10 +32,40 @@ enum {
 };
 
 enum {
+	TCMUR_DEV_LOCK_UNKNOWN,
 	TCMUR_DEV_LOCK_UNLOCKED,
 	TCMUR_DEV_LOCK_LOCKED,
+
+	/*
+	 * After the connection lost and before the first read fop
+	 * coming there was no any write fop for the device, then
+	 * the first read fop will set it and then reopen the device
+	 * without acquiring the lock later.
+	 */
+	TCMUR_DEV_LOCK_READ_REOPENING,
```

Comment:
> And one comment myself here about the current code, we still need to break the lock in read case and could skip acquiring it, because in case:
> 
> If current pathA lost the connection and if it have read fops, it will reopen the image without breaking and acquiring the lock, the pathB still could do read/write. There will be one problem if pathA want to do read-modiry-write things, the same offset maybe already updated by pathB just before pathA writing the modified data.
> 
> So we need to break the lock to blacklist pathB and to make sure the lock has no owner for every read in pathA, that means we must be sure the image is `readonly` temporily.

For the Read-Modify-Write or RMW case, from the linux scsi code, the old reopen code also has the same issue. Because  in kernel space the linux block/scsi code will just retry the cmds if its error == SAM_TS_BUSY. So for AA or shared device use case, if the initiator/apps want to do the write it must use the Revervation for locking or the atomic CAW/Compare-And-Write scsi cmd. 

The VMWare use case has no this issue.

Reply:
If this is true, then I will remove the READ_LOCKXXX flags.

Reply:
Right. Agree that you need reservations or use the atomic-test-set command.